lib/src/utf8.dart

```diff
@@ -31,15 +31,16 @@ class Utf8 extends Struct {
 
   /// Creates a [String] containing the characters UTF-8 encoded in [string].
   ///
-  /// The [string] must be a zero-terminated byte sequence of valid UTF-8
-  /// encodings of Unicode scalar values. A [FormatException] is thrown if the
-  /// input is malformed. See [Utf8Decoder] for details on decoding.
+  /// Either the [string] must be zero-terminated or its [length] -- the number
+  /// of bytes -- must be specified as a non-negative value. The byte sequence
+  /// must be valid UTF-8 encodings of Unicode scalar values. A [FormatException]
+  /// is thrown if the input is malformed. See [Utf8Decoder] for details on
+  /// decoding.
   ///
   /// Returns a Dart string containing the decoded code points.
-  static String fromUtf8(Pointer<Utf8> string) {
-    final int length = strlen(string);
-    return utf8.decode(Uint8List.view(
-        string.cast<Uint8>().asTypedList(length).buffer, 0, length));
+  static String fromUtf8(Pointer<Utf8> string, {int? length}) {
+    final len = length != null && length >= 0 ? length : strlen(string);
```

Comment:
Don't accept both negative values *and* `null` as as signals to use `strlen`. If you do pass an argument, it should be valid. I'd do:
```dart
if (length != null) {
  RangeError.checkNotNegative(length, 'length');
} else {
  length = strlen(strig);
}
```
and then use `length` below (which also avoids the `len` name, we discourage using abbreviations).

Reply:
Heh, you basically wrote the PR for me. Thanks for the help!